nil deref in etcdraft/multinode.go:234

The following test appears to have failed:

[#2374](https://circleci.com/gh/cockroachdb/cockroach/2374):

```
level="INFO" time="2015-05-12T21:03:18Z" file="multiraft.go" line="760" msg="node 300000003: group 1 raft ready"
level="INFO" time="2015-05-12T21:03:18Z" file="multiraft.go" line="765" msg="HardState updated: {Term:6 Vote:0 Commit:30 XXX_unrecognized:[]}"
level="INFO" time="2015-05-12T21:03:18Z" file="multiraft.go" line="774" msg="Snapshot updated: data:\"\\n<\\n\\030\\000\\377\\000\\377\\000\\377i\\t\\001res-\\010\\020\\353`se\\274\\nf\\000\\001\\022 \\022\\004\\010\\000\\020\\000\\030\\000 \\000(\\0002\\022\\n\\013\\202\\001\\010\\n\\006\\022\\004\\010\\000\\020\\004\\035:\\030\\234!\\nC\\n "
level="INFO" time="2015-05-12T21:03:18Z" file="multiraft.go" line="777" msg="Outgoing Message[0]: 12884901891->4294967297 MsgAppResp Term:6 Log:0/30"
level="WARNING" time="2015-05-12T21:03:18Z" file="multiraft.go" line="916" msg="node 100000001 failed to send message to 200000002: transport.go:169: unknown peer 200000002"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0xae5317]

goroutine 4930 [running]:
github.com/coreos/etcd/raft.(*multiNode).run(0xc208145a40)
	/go/src/github.com/coreos/etcd/raft/multinode.go:234 +0xc57
created by github.com/coreos/etcd/raft.StartMultiNode
	/go/src/github.com/coreos/etcd/raft/multinode.go:55 +0xb6

goroutine 1 [chan receive]:
testing.RunTests(0x10f6238, 0x15a9e20, 0xa5, 0xa5, 0xc207ffc501)
--
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:203 +0x2cb
github.com/cockroachdb/cockroach/util.func·001()
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:73 +0x51
created by github.com/cockroachdb/cockroach/util.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:74 +0xe3
FAIL	github.com/cockroachdb/cockroach/storage	3.803s
=== RUN TestBatchBasics
level="INFO" time="2015-05-12T21:03:56Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:03:56Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchBasics (0.00s)
=== RUN TestBatchGet
level="INFO" time="2015-05-12T21:03:56Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:03:56Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
level="INFO" time="2015-05-12T21:03:56Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:07:01Z" file="multiraft.go" line="777" msg="Outgoing Message[1]: 12884901891->4294967297 MsgAppResp Term:6 Log:0/56"
level="INFO" time="2015-05-12T21:07:01Z" file="multiraft.go" line="777" msg="Outgoing Message[2]: 12884901891->4294967297 MsgAppResp Term:6 Log:0/56"
level="WARNING" time="2015-05-12T21:07:01Z" file="multiraft.go" line="916" msg="node 200000002 failed to send message to 100000001: transport.go:169: unknown peer 100000001"
level="WARNING" time="2015-05-12T21:07:01Z" file="multiraft.go" line="916" msg="node 200000002 failed to send message to 100000001: transport.go:169: unknown peer 100000001"
level="WARNING" time="2015-05-12T21:07:01Z" file="multiraft.go" line="916" msg="node 300000003 failed to send message to 100000001: transport.go:169: unknown peer 100000001"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0xc4c5e9]

goroutine 2908 [running]:
github.com/coreos/etcd/raft.(*multiNode).run(0xc2084d34a0)
	/go/src/github.com/coreos/etcd/raft/multinode.go:234 +0xfa9
created by github.com/coreos/etcd/raft.StartMultiNode
	/go/src/github.com/coreos/etcd/raft/multinode.go:55 +0xde

goroutine 1 [chan receive]:
testing.RunTests(0x1392098, 0x18822a0, 0xa5, 0xa5, 0x1)
--
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1290 +0x10d9
github.com/cockroachdb/cockroach/util.func·001()
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:73 +0x80
created by github.com/cockroachdb/cockroach/util.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:74 +0x133
FAIL	github.com/cockroachdb/cockroach/storage	6.958s
=== RUN TestBatchBasics
level="INFO" time="2015-05-12T21:07:49Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:07:49Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchBasics (0.01s)
=== RUN TestBatchGet
level="INFO" time="2015-05-12T21:07:49Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:07:49Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
level="INFO" time="2015-05-12T21:07:49Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:07:01Z" file="multiraft.go" line="777" msg="Outgoing Message[1]: 12884901891->4294967297 MsgAppResp Term:6 Log:0/56"
level="INFO" time="2015-05-12T21:07:01Z" file="multiraft.go" line="777" msg="Outgoing Message[2]: 12884901891->4294967297 MsgAppResp Term:6 Log:0/56"
level="WARNING" time="2015-05-12T21:07:01Z" file="multiraft.go" line="916" msg="node 200000002 failed to send message to 100000001: transport.go:169: unknown peer 100000001"
level="WARNING" time="2015-05-12T21:07:01Z" file="multiraft.go" line="916" msg="node 200000002 failed to send message to 100000001: transport.go:169: unknown peer 100000001"
level="WARNING" time="2015-05-12T21:07:01Z" file="multiraft.go" line="916" msg="node 300000003 failed to send message to 100000001: transport.go:169: unknown peer 100000001"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0xc4c5e9]

goroutine 2908 [running]:
github.com/coreos/etcd/raft.(*multiNode).run(0xc2084d34a0)
	/go/src/github.com/coreos/etcd/raft/multinode.go:234 +0xfa9
created by github.com/coreos/etcd/raft.StartMultiNode
	/go/src/github.com/coreos/etcd/raft/multinode.go:55 +0xde

goroutine 1 [chan receive]:
testing.RunTests(0x1392098, 0x18822a0, 0xa5, 0xa5, 0x1)
--
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1290 +0x10d9
github.com/cockroachdb/cockroach/util.func·001()
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:73 +0x80
created by github.com/cockroachdb/cockroach/util.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:74 +0x133
FAIL	github.com/cockroachdb/cockroach/storage	6.958s
=== RUN TestBatchBasics
level="INFO" time="2015-05-12T21:07:49Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:07:49Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchBasics (0.01s)
=== RUN TestBatchGet
level="INFO" time="2015-05-12T21:07:49Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:07:49Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
level="INFO" time="2015-05-12T21:07:49Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:03:18Z" file="multiraft.go" line="760" msg="node 300000003: group 1 raft ready"
level="INFO" time="2015-05-12T21:03:18Z" file="multiraft.go" line="765" msg="HardState updated: {Term:6 Vote:0 Commit:30 XXX_unrecognized:[]}"
level="INFO" time="2015-05-12T21:03:18Z" file="multiraft.go" line="774" msg="Snapshot updated: data:\"\\n<\\n\\030\\000\\377\\000\\377\\000\\377i\\t\\001res-\\010\\020\\353`se\\274\\nf\\000\\001\\022 \\022\\004\\010\\000\\020\\000\\030\\000 \\000(\\0002\\022\\n\\013\\202\\001\\010\\n\\006\\022\\004\\010\\000\\020\\004\\035:\\030\\234!\\nC\\n "
level="INFO" time="2015-05-12T21:03:18Z" file="multiraft.go" line="777" msg="Outgoing Message[0]: 12884901891->4294967297 MsgAppResp Term:6 Log:0/30"
level="WARNING" time="2015-05-12T21:03:18Z" file="multiraft.go" line="916" msg="node 100000001 failed to send message to 200000002: transport.go:169: unknown peer 200000002"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0xae5317]

goroutine 4930 [running]:
github.com/coreos/etcd/raft.(*multiNode).run(0xc208145a40)
	/go/src/github.com/coreos/etcd/raft/multinode.go:234 +0xc57
created by github.com/coreos/etcd/raft.StartMultiNode
	/go/src/github.com/coreos/etcd/raft/multinode.go:55 +0xb6

goroutine 1 [chan receive]:
testing.RunTests(0x10f6238, 0x15a9e20, 0xa5, 0xa5, 0xc207ffc501)
--
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:203 +0x2cb
github.com/cockroachdb/cockroach/util.func·001()
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:73 +0x51
created by github.com/cockroachdb/cockroach/util.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:74 +0xe3
FAIL	github.com/cockroachdb/cockroach/storage	3.803s
=== RUN TestBatchBasics
level="INFO" time="2015-05-12T21:03:56Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:03:56Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchBasics (0.00s)
=== RUN TestBatchGet
level="INFO" time="2015-05-12T21:03:56Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"
level="INFO" time="2015-05-12T21:03:56Z" file="rocksdb.go" line="113" msg="closing in-memory rocksdb instance"
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
level="INFO" time="2015-05-12T21:03:56Z" file="rocksdb.go" line="88" msg="opening in-memory rocksdb instance"

```
Please assign, take a look and update the issue accordingly.

--- a/multiraft/multiraft.go
+++ b/multiraft/multiraft.go
@@ -388,6 +388,9 @@ type node struct {
 }
 
 func (n *node) registerGroup(groupID uint64) {
+	if groupID == noGroup {
+		panic("must not call registerGroup with noGroup")
+	}
 	n.groupIDs[groupID] = struct{}{}
 }
 
@@ -607,6 +610,9 @@ func (s *state) stop() {
 // are already registered, only the missing groups will be added in.
 func (s *state) addNode(nodeID proto.RaftNodeID, groupIDs ...uint64) error {
 	for _, groupID := range groupIDs {
+		if groupID == noGroup {
+			panic(fmt.Sprintf("attempt to add dummy group to node %d", nodeID))
+		}
 		if _, ok := s.groups[groupID]; !ok {
 			return util.Errorf("can not add invalid group %d to node %d",
 				groupID, nodeID)
@@ -903,13 +909,17 @@ func (s *state) sendMessage(groupID uint64, msg raftpb.Message) {
 	snapStatus := raft.SnapshotFinish
 	if err != nil {
 		log.Warningf("node %v failed to send message to %v: %s", s.nodeID, nodeID, err)
-		s.multiNode.ReportUnreachable(msg.To, groupID)
+		if groupID != noGroup {
+			s.multiNode.ReportUnreachable(msg.To, groupID)
+		}
 		snapStatus = raft.SnapshotFailure
 	}
 	if msg.Type == raftpb.MsgSnap {
 		// TODO(bdarnell): add an ack for snapshots and don't report status until
 		// ack, error, or timeout.
-		s.multiNode.ReportSnapshot(msg.To, groupID, snapStatus)
+		if groupID != noGroup {
+			s.multiNode.ReportSnapshot(msg.To, groupID, snapStatus)
+		}
 	}
 }
 

--- a/storage/client_raft_test.go
+++ b/storage/client_raft_test.go
@@ -572,7 +572,6 @@ func TestProgressWithDownNode(t *testing.T) {
 
 func TestReplicateAddAndRemove(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	t.Skip("currently disabled due to data race, see #1010")
 
 	// Run the test twice, once adding the replacement before removing
 	// the downed node, and once removing the downed node first.

--- a/storage/feed_test.go
+++ b/storage/feed_test.go
@@ -71,7 +71,6 @@ func waitForStopper(t testing.TB, stopper *util.Stopper) {
 
 func TestStoreEventFeed(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	t.Skip("currently disabled; see #1010")
 
 	// Construct a set of fake ranges to synthesize events correctly. They do
 	// not need to be added to a Store.

